jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+        return self.synced
+
+    def register_callbacks(self, callbacks, txinfo, cb_type="all"):
+        """ Register callbacks that will be called by the
+        transaction monitor loop, on transactions stored under
+        our wallet label (see WalletService.get_wallet_name()).
+        Callback arguments are currently (txd, txid) and return
+        is boolean, except "confirmed" callbacks which have
+        arguments (txd, txid, confirmations).
+        Note that callbacks MUST correctly return True if they
+        recognized the transaction and processed it, and False
+        if not. The True return value will be used to remove
+        the callback from the list.
+        Arguments:
+        `callbacks` - a list of functions with signature as above
+        and return type boolean.
+        `txinfo` - either a txid expected for the transaction, if
+        known, or a tuple of the ordered output set, of the form
+        (('script': script), ('value': value), ..). This can be
+        constructed from jmbitcoin.deserialize output, key "outs",
+        using tuple(). See WalletService.transaction_monitor().
+        `cb_type` - must be one of "all", "unconfirmed", "confirmed";
+        the first type will be called back once for every new
+        transaction, the second only once when the number of
+        confirmations is 0, and the third only once when the number
+        of confirmations is > 0.
+        """
+        if cb_type == "all":
+            # note that in this case, txid is ignored.
+            self.callbacks["all"].extend(callbacks)
+        elif cb_type in ["unconfirmed", "confirmed"]:
+            if txinfo not in self.callbacks[cb_type].keys():
+                self.callbacks[cb_type][txinfo] = []
+            self.callbacks[cb_type][txinfo].extend(callbacks)
+        else:
+            assert False, "Invalid argument: " + cb_type
+
+
+    def start_wallet_monitoring(self, syncresult):
+        """ Once the initialization of the service
+        (currently, means: wallet sync) is complete,
+        we start the main monitoring jobs of the
+        wallet service (currently, means: monitoring
+        all new transactions on the blockchain that
+        are recognised as belonging to the Bitcoin
+        Core wallet with the JM wallet's label).
+        """
+        if not syncresult:
+            jlog.error("Failed to sync the bitcoin wallet. Shutting down.")
+            reactor.stop()
+        jlog.info("Starting transaction monitor in walletservice")
+        self.monitor_loop = task.LoopingCall(
+            self.transaction_monitor)
+        self.monitor_loop.start(5.0)
+
+    def transaction_monitor(self):
+        """Keeps track of any changes in the wallet (new transactions).
+        Intended to be run as a twisted task.LoopingCall so that this
+        Service is constantly in near-realtime sync with the blockchain.
+        """
+        txlist = self.bci.list_transactions(100)
```

Comment:
This looks like it's just waiting to cause trouble for someone. Only ever the 100 newest transactions are considered. bitcoind's `listtransactions` supports a `skip` parameter which could be used (which would simply be `len(self.old_txs)` if I'm not missing anything)

Reply:
Note that we had this same logic and magic number already in the previous version (just, we were doing it per join, which is kinda yuck).

I agree that it's not a good thing, but the likelihood of it causing something to be missed is negligible (100+ transaction *updates* in parallel; something else will completely break before this), and the consequences are at worst a wallet not knowing about a tx, not a funds loss risk. My feeling is to leave it as-is for this PR, and to open an Issue to fix it to be properly generic.

Reply:
I'm ok with leaving an issue for that.